### PR TITLE
OpTest enhancement: unify forward backward logics; bug fix; support in-place operator test

### LINKED ISSUE #5842: Support "In place" computation in operator unit test
`OpTest` in `op_test.py` is currently the base class of all operator unit tests, and it doesn't provide any method to specify certain input tensor and output tensor share the same memory, i.e. the operator does in place computation.

We should enhance this to support easier unit test of operators like batch normalization.

### PR DESCRIPTION
**Pull request by zhouxiao-coder**
resolve #5842 

1. As described in issue #5842, I was trying to add a small feature to OpTest class to support convenient unit test for in-place computation operators, which is supported by the `in_place_map` attribute added in base class `OpTest`.

2. After reading the original implementation, I find several places that can be improved and start to rewrite some of the methods, which mainly to improve the following aspects:

### Unify forward/backward code
In the old implementation, forward test and backward test are implemented in completely different logics, probably due to historical reasons. They use different ways to create operators, to create variable descriptions, feed variables, and execute the program. Check the "#FIXME" comment in the old code.

In this PR, both forward and backward use similar process to do computation:
```python
#1. initialize program, global block
self._init_program()

#2. compile the operator, return input and output descriptions
op, input_var_descs, output_var_descs = self._compile_op(place)

#3. prepare input and output of the operator
feed_map = self._feed_vars(input_var_descs, self.inputs, place)
fetch_list = self._get_fetch_list(output_var_descs, place, filter=self.outputs.keys())

#4. execute
outs = self._execute(feed_map, fetch_list, place)
```

### Scope and Block
The old code uses both `block` and `scope` to manage variables. I find it quite confusing, so I completely remove `scope` related code and replace it with `block` methods. 

I'm actually a little fuzzy about the `scope` and `block` concept in PaddlePaddle, even after reading all the design docs in current repo, so **Please** help to make sure this is ok. I ran all the unit tests, it seems alright.

### New batch_norm_op unit test
Due to lack of in-place computation support, the original batch_norm_op unit test rewrites most of the logic in the `OpTest` base class. Now a simpler version is possible after the enhancement.

### CODE REVIEW

**Comment by zhouxiao-coder** [comment]
Currently, CI is blocked by batch_norm_op test failure on GPU case. 
It is actually a bug in the c++ batch_norm_op implementation: 
- the GPU version uses CUDNN batch normalization implementation, which adjusts batch variance with Bessel's correction by default; 
- the CPU version has no such adjustment, so the running variance output is inconsistent between two modes.
I'm fixing this in another PR #5845

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献代码。由于Paddle V1/V2版本已不再维护，相关代码也已从develop分支上删除，因此关闭您的PR，欢迎您向Paddle最新版-Fluid贡献代码。
Thanks for contributing to PaddlePaddle! Since V1/V2 will not be maintained anymore, and related codes have been deleted from develop branch as well, we close this PR. Welcome to contribute to Fluid——the latest version of PaddlePaddle.
